R/ResultsDataModelQueries.R

```diff
@@ -0,0 +1,294 @@
+#' @export
+createDatabaseDataSource <- function(connection,
+                                     resultsDatabaseSchema,
+                                     vocabularyDatabaseSchema = resultsDatabaseSchema) {
+  return(
+    list(
+      connection = connection,
+      resultsDatabaseSchema = resultsDatabaseSchema,
+      vocabularyDatabaseSchema = vocabularyDatabaseSchema
+    )
+  )
+}
+
+
+#' Return a file data source object
+#'
+#' @description
+#' Given a premerged file (an .RData/rds object the output
+#' of \code{CohortDiagnostics::preMergeDiagnosticsFiles} reads the object into
+#' memory and makes it available for query.
+#'
+#' @param premergedDataFile  an .RData/rds object the output
+#'                           of \code{CohortDiagnostics::preMergeDiagnosticsFiles}
+#'
+#' @return
+#' R environment containing data conforming to Cohort Diagnostics results data model specifications.
+#' @export
+createFileDataSource <-
+  function(premergedDataFile, envir = new.env()) {
+    load(premergedDataFile, envir = envir)
+    return(envir)
+  }
+
+
+# private function - not exported
+# this function supports query with connection that is either pool or not pool. It can also establish a connection
+# when connection details is provided
+renderTranslateQuerySql <-
+  function(sql,
+           connection = NULL,
+           connectionDetails = NULL,
+           ...,
+           snakeCaseToCamelCase = FALSE) {
+    # Set up connection to server ----------------------------------------------------
+    if (is.null(connection)) {
+      if (!is.null(connectionDetails)) {
+        writeLines("Connecting to database using provided connection details.")
+        connection <- DatabaseConnector::connect(connectionDetails)
+        on.exit(DatabaseConnector::disconnect(connection))
+      } else {
+        stop("No connection or connectionDetails provided.")
+      }
+    }
+    
+    if (methods::is(connection, "Pool")) {
+      # Connection pool is used by Shiny app, which always uses PostgreSQL:
+      sql <- SqlRender::render(sql, ...)
+      sql <- SqlRender::translate(sql, targetDialect = "postgresql")
+      
+      tryCatch({
+        data <- DatabaseConnector::dbGetQuery(connection, sql)
+      }, error = function(err) {
+        writeLines(sql)
+        stop(err)
+      })
+      if (snakeCaseToCamelCase) {
+        colnames(data) <- SqlRender::snakeCaseToCamelCase(colnames(data))
+      }
+      return(data %>% dplyr::tibble())
+    } else {
+      if (is.null(attr(connection, "dbms"))) {
+        stop("dbms not provided. Unable to translate query.")
+      }
+      return(
+        DatabaseConnector::renderTranslateQuerySql(
+          connection = connection,
+          sql = sql,
+          ...,
+          snakeCaseToCamelCase = snakeCaseToCamelCase
+        ) %>% dplyr::tibble()
+      )
+    }
+  }
+
+# private function - not exported
+getDataForDatabaseIdsCohortIds <- function(dataSource,
```

Comment:
This function has a lot of flow control so really needs some careful unit tests.